### PR TITLE
SG-456 Cache bucket metadata

### PR DESCRIPTION
**Pull request by lech-lorens-epam**
## Description
This set of commits implements caching of bucket metadata.

## Motivation and Context
The goal of this change is to improve the performance of MinIO on PanFS.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/panfs_config_cache.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+// CachingConfigAccessor can be used with the config operations implemented in
+// config-common.go (readConfig(), deleteConfig(), saveConfig()) to cache the
+// most frequently accessed data.
+type CachingConfigAccessor struct {
+	cache       *cache.Cache
+	rawAccessor configAccessor
+}
+
+func newCachingConfigAccessor(raw configAccessor) *CachingConfigAccessor {
```
Comment (yauhen-matsiushonak) [inline cmd/panfs_config_cache.go]: I'm not sure that we need eviction policy for bucket metadata at all. We have to know about all buckets and when the list is populated - we can fit it in memory. We will not have too many of them. Instead, I'd make sure that we invalidate cached metadata each time when we receive signal from mcproxy (there is signal handler in our fork of MinIO that reloads IAM cache).

Comment (yauhen-matsiushonak) [reply cmd/panfs_config_cache.go]: Just as an idea, maybe it worth to adopt BucketMetadataSys (global instance is a globalBucketMetadataSys) described in /cmd/bucket-metadata-sys.go for our gateway mode. Other modes use it as in-memory cache of bucket metadata. Maybe we can just add Load/List functions to it?

File: cmd/bucket_metadata_cache.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"errors"
+	"sync"
+)
+
+var errCacheEntryNotFound = errors.New("Entry not found in cache")
+
+type BucketMetadataCache struct {
+	lock     sync.Mutex
+	metadata map[string]BucketMetadata
```
Comment (yauhen-matsiushonak) [inline cmd/bucket_metadata_cache.go:12]: As we will be calling 'Get()' in 99% of cases, can we use rwlock instead of mutex?

**Review by yauhen-matsiushonak — APPROVED** [review]
overall change looks good for me, and I reviewed it in process of development. Please address commend about rwlock if you think it is reasonable.

File: cmd/bucket_metadata_cache.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+// unlock unlocks cache locked for writing
+func (cache *BucketMetadataCache) unlock() {
+	cache.mutex.Unlock()
+}
+
+// Get fetches the given value from the cache
+func (cache *BucketMetadataCache) Get(bucket string) (BucketMetadata, error) {
+	if cache == nil {
```
Comment (dmitrizpanasas) [inline cmd/bucket_metadata_cache.go:44]: It seems that this check is redundant - if go comes here then it means that `cache` cannot be `nil`.

Have you faced with some kind of errors related to this check?

Comment (lech-lorens-epam) [reply cmd/bucket_metadata_cache.go:44]: In go you can call a method using a nil pointer.
And, yes, I got crashes in some unit tests.

**Review by dmitrizpanasas — APPROVED** [review]

**Comment by lech-lorens-epam** [comment]
Merged manually.

**Pull request closed without merging**